Class를 통한 PlanView Refactoring

seoko97 opened the pull request:
* Closes #61

## ✨ **구현 기능 명세**
- 기존 `getCalendarPlans.ts` 제거 및 `DaysPlanManager` 생성
- `DaysPlanManager`에 저장된 `viewInfo`에 따라 로직 변경

## 🎁 **주목할 점** 
- DaysPlanManager`는 일정 범위를 입력받아 실제 렌더링될 일정에 정보들을 반환합니다.
- getViewPlan를 실행하여 다음과 같은 정보를 반환합니다.

```ts
  getViewPlan(plan: Plan): IDayViewInfo | null {
   ...
    return {
      // plan id
      id: plan.id,
      // view의 시작 요일
      start: viewStart.day(),
      // view의 인덱스
      index: 0,
      // view의 시작날짜
      st: viewStart,
      // view의 종료날짜
      et: viewEnd,
      // view의 길이
      term,
    };
  }
```

- 이후 view의 시작 날짜와 길이에 따라 순서를 결정합니다.

```ts
  setIndex(viewPlans: Map<Plan['id'], IDayViewInfo>) {
    const rangeTerm = Math.abs(this.startDate.diff(this.endDate, 'd')) + 1;
    const rangeIndex: number[][] = Array.from({ length: rangeTerm }, () => []);

    viewPlans.forEach((planView) => {
      let index = 0;

      const { start, term } = planView;

      for (let i = 0; i < term; i++) {
        const indexInfo = rangeIndex[start + i];

        while (indexInfo[index] !== planView.id) {
          if (!indexInfo[index]) {
            indexInfo[index] = planView.id;
            planView.index = index;
          } else {
            index++;
          }
        }
      }
    });

    // 달력 주차의 height를 설정하기 위해 각 일자에 저장된 id 배열을 저장합니다.
    this.daysIndex = rangeIndex;
  }
```

- 다음과 같은 로직을 통해 기존 한달 단위의 일정 정보를 입력한 일자 단위로 일정 정보를 확인할 수 있습니다.\

## 😭 **어려웠던 점**
- `Plan`를 생성하면서 기존 `iViewPlan`을 사용하던 컴포넌트를 모두 변경하였습니다.

seoko97 committed:
feat: DaysPlanManager 생성 (#61)

seoko97 committed:
feat: 사용하지 않는 파일 제거

seoko97 committed:
a

seoko97 committed:
Merge branch 'dev' into refactor/days_paln_manager#61

seoko97 committed:
feat: 특정 일자의 일정 개수 저장 (#61)

seoko97 committed:
feat: planManager에 저장된 plan을 보여주는 CalendarLayer 생성 (#61)

seoko97 committed:
feat: 선택된 월을 벗어나는 일자를 위한 CalendarOverlay 컴포넌트 추가

seoko97 committed:
feat: 달력의 주차를 표시하는 CalendarWeek 컴포넌트 추가

seoko97 committed:
feat: PlanManager를 통한 컴포넌트 분리 (#61)

seoko97 committed:
feat: component 최적화 (#61)

seoko97 committed:
feat: 달력의 주차 범위에 따라 PlanManager을 반환하는 함수 구현 (#61)

seoko97 committed:
feat: refactor을 통한 getCalendarPlans 제거 (#61)

pyo-sh commented on src/components/calendars/large/CalendarView.tsx:
```diff
@@ -1,30 +1,41 @@
-  const calendarInfos = getCalendarInfo({ year, month, day });
-  const dates = getStartAndEndDateInMonth(calendarInfos);
-  const selectedPlanArgs = focusedPlan ? [focusedPlan] : [];
+  const calendarInfos = useMemo(
+    () => getCalendarInfo({ year, month, day }),
+    [year, month, day],
+  );
```
이 부분의 useMemo는 year, month, day 값들을 바로 사용하는 것이라서 필요없는 것 아닌가요?

seoko97 replied:
`calendarInfos`는 해당 달의 달력에 대한 기본적인 정보를 반환합니다. 만약 `useMemo`를 사용하지 않는다면 `CalendarView`의 state가 변경되는 상황에서 쓸데없는 렌더링이 발생하여 useMemo의 dependency에 일자를 주입해줬습니다.

pyo-sh replied:
제가 리뷰남길 당시 useMemo의 결과 값이 year, month, day 인 줄 착각하고 있었네요😢 죄송합니다.
View 안에 Dragging, Hover 등 상태 변경이 많아 계산을 줄이기 위함이였군요!

pyo-sh commented on src/utils/plan/getDaysPlanManager.ts:
```diff
@@ -0,0 +1,43 @@
+  const arr: DaysPlanManager[] = [];
+
+  const planClass = plans
+    .filter((plan) => plan.type === 'task')
```
타입이 `task` 일 경우에만 달력에 노출되는 것인가요?
함께 코딩할 때 확실히 결정한 사항이 없었던 것 같아 질문합니다.

seoko97 replied:
일정에 타입에 따라 렌더링하는 방법이 달라 분리했습니다. `type: task`일 경우 일정에 길이에 따라 주차에 걸쳐 일정이 렌더링 되어야 하지만, 그 외의 경우는 특정 일자 하루에 일정이 존재해야하며 일정의 style, 일정이 렌더링될 위치가 다르기 때문에 필터링해주었습니다.

pyo-sh replied:
이해했습니다. `task`와 이외 타입들을 분리해서 렌더링 할 예정인 것이군요

pyo-sh commented on src/plan/DaysPlanManager.ts:
```diff
@@ -0,0 +1,114 @@
+class DaysPlanManager extends PlanManager<IDayViewInfo> {
+  startDate: Moment;
+  endDate: Moment;
+  daysIndex: number[][] = [];
```
`daysIndex`를 Class 속성으로 둔 이유가 있나요?

저는 `daysIndex`가 DayPlan들의 각 순서를 정하기 위한 배열인 것으로 이해했는데
각 순서를 정하고 나서 그 순위만 PlanInfo에 저장해준다면 해당 배열은 사용하지 않는 것 아닌가요?

seoko97 replied:
달력에서 height를 지정해주기 위해 각 일자에 몇 개의 일정이 존재하는지 필요합니다.

일정에 대한 순서만 저장할 경우 특정일자에 어떤 일정이 존재하는지 다시 계산하는 로직이 필요하기때문에 따로  저장해주었습니다

pyo-sh replied:
> 달력에서 height를 지정해주기 위해 각 일자에 몇 개의 일정이 존재하는지 필요합니다.

달력 각 일자의 일정 개수가 height를 지정한다는 말씀이시군요.
위의 `DaysPlanManager`가 `type: task` 만을 다룬다는 것을 제가 모르고있어 이해가 되지 않았던 것 같습니다!

jintak0401 commented on src/components/calendars/large/CalendarLayer.tsx:
```diff
@@ -0,0 +1,85 @@
+
+  const onMouseLeave = useCallback(() => {
+    clearHoveredPlan();
+  }, []);
```
`plans` 에 많은 일정이 담겨있을 때, 일정들을 대상으로 마우스 이벤트가 일어나면(`onMouseEnter`, `onMouseDown`, `onMouseLeave` 등) 리렌더링시 다시 함수를 만들지 않도록 최적화를 위해서 사용하신 것 같습니다.


```tsx
<DayPlan
    onMouseLeave={clearHoveredPlan}
    ...
/>
```

그런데 `onMouseLeave` 같은 경우는 `useHoveredPlan` 에서 제공하는 함수만 사용하고 있으니, 위와 같이 작성하면 안될까요?

seoko97 replied:
확인했습니다. `zustand`에서 제공하는 함수는 렌더링 상황에서 재생성이 되지 않는군요... 하나 배웠습니다.

seoko97 committed:
feat: getViewPlan 함수 null type 제거 (#61, #68)

seoko97 committed:
feat: 주석 제거 (#68)

seoko97 committed:
feat: 사용하지 않는 함수 제거 (#68)

jintak0401 commented on src/components/calendars/large/CalendarView.tsx:
```diff
@@ -1,30 +1,41 @@
+
+  const focusedPlanManager = useMemo(
+    () => getDaysPlanManager(focusedPlanArgs, calendarInfos),
+    [focusedPlanArgs, calendarInfos],
```
`focusedPlanArgs` 의 경우 리렌더링시 매번 새로운 배열을 가지는 변수인 것 같습니다. 그렇다면 deps 배열에 `focusedPlanArgs`를 넣어주는 것은 `useMemo`의 이점을 살리지 못하는 것 같습니다.

```suggestion
 const focusedPlanManager = useMemo(
    () => getDaysPlanManager(focusedPlan ? [new Plan(focusedPlan)] : [], calendarInfos),
    [focusedPlan, calendarInfos],
)
```

로 해야 `useMemo`의 이점을 살릴 수 있을 것 같습니다.

seoko97 replied:
확인했습니다. 렌더링 상황에서 `focusedPlanArgs`가 새로운 주소값을 가지는 것 같습니다. 기존 코드에서 조금씩 수정하다보니 이런 실수를 했네요..

jintak0401 commented on src/components/calendars/large/CalendarWeek.tsx:
```diff
@@ -0,0 +1,45 @@
+          dateInfo={dateInfo}
+          format={dateInfo.format}
+          isSelected={dateInfo.day === day && dateInfo.isInMonth}
+          onClickDayNumber={onChangeStoreDate}
```
`zustand`의 store에 있는 함수인데 굳이 props로 넘겨주어야 하나요?

seoko97 replied:
`useDateState`의 `day` state를 `CalendarWeek`에서 사용하면서 함께 불러와 넘겨주었습니다.

이전 로직에서 무분별하게 `zustand`의 store를 불러와 사용하였는데 리렌더링에 쓴맛을 느껴 해당 코드처럼 작성했습니다..

pyo-sh replied:
```typescript
const onChangeStoreDate = useDateState((state) => state.onChangeStoreDate);
```
처럼 사용한다면 상태가 바뀌어도 함수가 변경되지 않는다면 리렌더링이 일어나지 않을 것 같습니다!

seoko97 replied:
> ```ts
> const onChangeStoreDate = useDateState((state) => state.onChangeStoreDate);
> ```
> 
> 처럼 사용한다면 상태가 바껴도 함수가 변경되지 않는다면 리렌더링이 일어나지 않을 것 같습니다!

redux `useSelector`같이 사용할 수 도 있군요! 확인했습니다.ᐟ

seoko97 committed:
feat: useMemo dependency 변경 (#68)
- 기존 외부에서 변수에 주소값을 할당하여 useMemo 내부에서 사용
- 리렌더링 상황시 항상 변수의 주소값이 바뀌어 useMemo가 정상적으로 적용되지 않음
- 해당 변수를 제거하고 useMemo 내부에서 판단

seoko97 committed:
feat: onMouseLeave 제거 (#68)
- zustand에서 생성한 store 함수는 리렌더링 상황에서 재생성되지 않음
- 기존 함수를 제거하고 store 함수를 props로 넘겨줌

jintak0401 reviewed (approved):
고생하셨습니다!

seoko97 committed:
feat: props 제거 및 사용하는 함수에서 해당 함수 호출하는 로직으로 변경 (#68)

pyo-sh reviewed (approved):
리팩토링에 고생이 많으십니다..!
